src/Managed.hs

```diff
@@ -0,0 +1,35 @@
+module Managed where
+
+import Lookup
+import Obj
+import Types
+
+-- | Find out if a type is "external", meaning it is not defined by the user
+--   in this program but instead imported from another C library or similar.
+-- NOTE: Quite possibly this function should be removed and we should rely on 'isManaged' instead?
+isExternalType :: TypeEnv -> Ty -> Bool
+isExternalType typeEnv (PointerTy p) =
+  isExternalType typeEnv p
+isExternalType typeEnv (StructTy (ConcreteNameTy name) _) =
+  case lookupBinder (SymPath [] name) (getTypeEnv typeEnv) of
+    Just (Binder _ (XObj (Lst (XObj (ExternalType _) _ _ : _)) _ _)) -> True
+    Just _ -> False
+    Nothing -> False
+isExternalType _ _ =
+  False
+
+-- | Is this type managed - does it need to be freed?
+isManaged :: TypeEnv -> Ty -> Bool
+isManaged typeEnv (StructTy (ConcreteNameTy name) _) =
+  (name == "Array") || (name == "StaticArray") || (name == "Dictionary")
+    || ( case lookupBinder (SymPath [] name) (getTypeEnv typeEnv) of
+           Just (Binder _ (XObj (Lst (XObj (ExternalType _) _ _ : _)) _ _)) -> False
+           Just (Binder _ (XObj (Lst (XObj (Deftype _) _ _ : _)) _ _)) -> True
+           Just (Binder _ (XObj (Lst (XObj (DefSumtype _) _ _ : _)) _ _)) -> True
+           Just (Binder _ (XObj wrong _ _)) -> error ("Invalid XObj in type env: " ++ show wrong)
```

Comment:
Probably it should be caller's responsibility to handle these last two `error`ing cases? If so, I think the only check that's relevant is the `ExternalType` check, since we are probably ok assuming the `StructTy` passed to this is a valid type in the type env (if not, the error is on the caller). Then we could redefine the struct case:

```haskell
isManaged typeEnv (StructTy (ConcreteNameTy name) _) =
  (name == "Array") || (name == "StaticArray") || (name == "Dictionary")
  || ( case lookupBinder (SymPath [] name) (getTypeEnv typeEnv) of
         Just (Binder _ (XObj (Lst (XObj (ExternalType _) _ _ : _)) _ _)) -> False
         _ -> True -- all other structTys are managed
```

This would make it a pure function proper and push our errors upstream.

Reply:
Or better yet:

```haskell
( case lookupBinder (SymPath [] name) (getTypeEnv typeEnv) of
         Just (Binder _ (XObj (Lst (XObj (ExternalType _) _ _ : _)) _ _)) -> False
         _ -> True -- all other structTys are managed
```

(since the last case covers the specific matches on names)

Reply:
I'll rewrite this function soon anyways.